### DISCUSSION TITLE
Ambiguity of Property wrappers and Macros

### DISCUSSION THREAD

**Discussion by DandyLyons** [pointfreeco/swift-composable-architecture#2861] [General]
One problem I have with Swift macros is that they look exactly the same as property wrappers at the call site. For example: 
@Reducer, @ObservableState, @CasePathable are macros. Whereas
@Presents, and @Shared are property wrappers. 

Both macros and property wrappers write boiler plate code for us but they do so in very different ways, and require us to interact with them differently. For example: 
- we can “Expand” macros. We can’t expand property wrappers. 
- Property wrappers require us to use wrappedValue and projectedValue as well as $property and _property syntax. 

This certainly isn’t TCA’s fault, but I do think it complicates learning TCA, especially for beginners and those transitioning from pre-macro TCA. The user has to create a mental model of which is a macro and which is a property wrapper, and they have to understand the under the hood subtleties of how that effects their call site. 

For example, it can be very frustrating to have an entire project fail to compile simply because I forgot to include a `$` symbol. And it's not obvious at the call site why one place needs a `$` symbol and the other doesn't. For example: 
```
.ifLet(\.$child, action: \.child) {
      ChildFeature()
}
```

This isn't so much a criticism of TCA, nor a recommendation of something to change. I'm more wondering how we can help in making this mental model easier to learn and retain.

**Comment by stephencelis** [comment]
@DandyLyons You go out of your way of saying that this isn't TCA's fault, but I'd like to go further and point out that this same thing applies to modern SwiftUI: you have a ton of property wrappers (`@State`, `@Binding`, `@FocusState`, `@ObservedObject`, etc.), but now you have `@Observable`, `@Query`, and other macros being introduced, and some of these macros are _also_ property wrappers and thus Xcode won't expand them.

So I'm not sure TCA can do much when it comes to property wrappers vs. macros other than wait for Apple's vanilla tooling to be better about it. But I also think that ideally a user doesn't have to really think about whether something is a macro or a property wrapper. They can both achieve the same things and the same syntax. In fact, `@Presents` _is_ a macro even though it masquerades as a property wrapper with `_` and `$` syntax.

One thing we _can_ do is simplify things in the future, and our internal roadmap aims to simplify the number of `@`s (and `$`s) typically seen in a code base. We've taken some steps already with observation removing the need for `@BindingState`. It will take a 2.0 with breaking changes, and some time, but eventually we also want to:

- Automatically apply `@ObservableState` via the `@Reducer` macro
- Remove `@Presents` in favor of simple optionals, which also means no more `$` projection for the `ifLet` reducer

This would mean `@Reducer` is the most common `@` seen in a TCA code base. And after that, the occasional `@Shared` property wrapper and non-action `@CasePathable` enum.

For example, a reducer without shared state will only see `@Reducer`:

```swift
@Reducer
struct Feature {
  struct State: Equatable {
    var child: Child.State?
  }
  …
  var body: some ReducerOf<Self> {
    Reduce { state, action in
      …
    }
    .ifLet(\.child, action: \.child) {
      Child()
    }
  }
}
```

**Comment by DandyLyons** [comment]
> You go out of your way of saying that this isn't TCA's fault, but I'd like to go further and point out that this same thing applies to modern SwiftUI: you have a ton of property wrappers (@State, @Binding, @FocusState, @ObservedObject, etc.), but now you have @Observable, @Query, and other macros being introduced, and some of these macros are also property wrappers and thus Xcode won't expand them.

Agreed. It really is a language level problem in my opinion. I really wish they chose a different symbol than `@` for macros, but it's too late for that. In the meanwhile, I think we need to wait for Apple to mitigate this problem.

**Reply by jegnux** [reply]
@DandyLyons i don't agree with that. @ is not to denote the implementation form of an underlying feature (property wrapper or macro) but to denote that it adds a new behavior to the element, wether this behavior is a macro (@Observable), a property wrapper (@Binding) a result builder (@ViewBuilder) or language behavior (@Sendable, @escaping, @available, etc)

On the other side # things expand in place (#file, #line, #available, and # macros)

But one thing I agree is that Xcode could expand more things. Property wrappers could theoretically be expanded as well, and we could see how $someVar and _someVar and someVar are backed (even if it's always the same way with a property wrapper)

**Reply by DandyLyons** [reply]
@jegnux I suppose you're right. But it doesn't change the fact that PW and macros have very different semantics and behavior that is not understandable unless you know which one it is. 

Agreed. Perhaps if PW were expandable this would be less of an issue.
